### ISSUE TITLE
sql: SHOW GRANTS on empty db does not show grant on db until new table is created.

### ISSUE THREAD

**Issue by varunturlapati** [cockroachdb/cockroach#32811]
**Describe the problem**

After freshly installing CockroachDB and creating 2 user, I tried to grant privileges to a user. The command doesn't fail but the desired outcome is not achieved. Upon looking into all the `grants` for the user it shows blank.

**To Reproduce**
`
#cockroach sql --insecure --user=root
#Welcome to the cockroach SQL interface.
#All statements must be terminated by a semicolon.
#To exit: CTRL + D.
#
#Server version: CockroachDB CCL v2.1.1 (x86_64-unknown-linux-gnu, built 2018/11/19 18:24:21, go1.10.3) (same version as client)
#Cluster ID: e76fcc16-02a9-41d6-90a6-5679d62302dc
#
#Enter \? for a brief introduction.
#
root@:26257/defaultdb> show databases;
  database_name
+---------------+
  apimgtdb
  defaultdb
  postgres
  regdb
  statdb
  system
  userdb
(7 rows)

Time: 2.1338ms
root@:26257/defaultdb> show users;
   user_name
+-------------+
  cmu
  root
  wso
(3 rows)

Time: 1.847701ms
root@:26257/defaultdb> GRANT SELECT ON DATABASE apimgtdb to wso;
GRANT

Time: 5.316302ms

root@:26257/defaultdb> show grants for wso;
  database_name | schema_name | table_name | grantee | privilege_type
+---------------+-------------+------------+---------+----------------+
(0 rows)

Time: 4.539601ms

`

**Expected behavior**
That user `wso` should have privileges

**Environment:**
# cockroach version
Build Tag:    v2.1.1
Build Time:   2018/11/19 18:24:21
Distribution: CCL
Platform:     linux amd64 (x86_64-unknown-linux-gnu)
Go Version:   go1.10.3
C Compiler:   gcc 6.3.0
Build SHA-1:  4523f9819d2dde9b054fc8faf1568e107f130986
Build Type:   release

OS: Centos7

**Comment by knz** [comment]
@petermattis @jordanlewis here's another area of the product where I do not exactly know how to triage. Do you have suggestions?

**Comment by lanzao** [comment]
@varunturlapati  Whether `apimgtdb` has any table?

I have tried to do according to your desc and found that if I grant privileges to a user on a database which doesn't have any table ,`show grants for xxx` will not show the privilege.

Once you create a table on database, it will work correctly.

**Comment by knz** [comment]
@lanzao good call on noticing that the problem is specific to databases without any tables. This narrows down the situation for us. Thank you.

**Comment by knz** [comment]
This issue still exists in v21.1.
Repro on an empty cluster:

```
CREATE USER foo;
GRANT SELECT ON DATABASE defaultdb TO foo;
SHOW GRANTS FOR defaultdb;
```

cc @rafiss for re-triage

**Comment by rafiss** [comment]
It seems like SHOW GRANTS does work fine on an empty database

```
root@:26257/defaultdb> create user wso;
CREATE ROLE

root@:26257/defaultdb> create database apimgtdb;
CREATE DATABASE

root@:26257/defaultdb> GRANT CONNECT ON DATABASE apimgtdb TO wso;
GRANT

root@:26257/defaultdb> show grants on database apimgtdb for wso;
  database_name | grantee | privilege_type
----------------+---------+-----------------
  apimgtdb      | wso     | CONNECT
(1 row)

root@:26257/defaultdb> use apimgtdb;
SET

root@:26257/apimgtdb> show grants for wso;
  database_name | schema_name | relation_name | grantee | privilege_type
----------------+-------------+---------------+---------+-----------------
  apimgtdb      | NULL        | NULL          | wso     | CONNECT
```

I think some of the confusion here is that `SHOW GRANTS` only shows grants from the current session database, unless the `FOR DATABASE <db>` clause is added. Also, maybe this didn't used to work correctly, but it is fixed now.

---

Separately, the previous comment in this thread was using the `SHOW GRANTS FOR <username>` syntax, but a database name was used instead of a username. Instead of showing a "username does not exist" error, the command returns 0 rows.

Let's rename this issue so that instead it fixes this confusion.

The following should return an error message instead of succeeding with 0 rows.

```
root@:26257/defaultdb> show grants for non_existent_user;
SHOW GRANTS 0
```

**Comment by jazoom** [comment]
> It seems like SHOW GRANTS does work fine on an empty database
> 
> ```
> root@:26257/defaultdb> create user wso;
> CREATE ROLE
> 
> root@:26257/defaultdb> create database apimgtdb;
> CREATE DATABASE
> 
> root@:26257/defaultdb> GRANT CONNECT ON DATABASE apimgtdb TO wso;
> GRANT
> 
> root@:26257/defaultdb> show grants on database apimgtdb for wso;
>   database_name | grantee | privilege_type
> ----------------+---------+-----------------
>   apimgtdb      | wso     | CONNECT
> (1 row)
> 
> root@:26257/defaultdb> use apimgtdb;
> SET
> 
> root@:26257/apimgtdb> show grants for wso;
>   database_name | schema_name | relation_name | grantee | privilege_type
> ----------------+-------------+---------------+---------+-----------------
>   apimgtdb      | NULL        | NULL          | wso     | CONNECT
> ```
> 
> I think some of the confusion here is that `SHOW GRANTS` only shows grants from the current session database, unless the `FOR DATABASE <db>` clause is added. Also, maybe this didn't used to work correctly, but it is fixed now.
> 
> Separately, the previous comment in this thread was using the `SHOW GRANTS FOR <username>` syntax, but a database name was used instead of a username. Instead of showing a "username does not exist" error, the command returns 0 rows.
> 
> Let's rename this issue so that instead it fixes this confusion.
> 
> The following should return an error message instead of succeeding with 0 rows.
> 
> ```
> root@:26257/defaultdb> show grants for non_existent_user;
> SHOW GRANTS 0
> ```

This doesn't appear to be fixed in 21.2.5.

I was just following https://www.cockroachlabs.com/docs/stable/authorization.html to set up grants and I couldn't figure out why `SHOW GRANTS;` wasn't returning the grants I had just applied.

For me, following the instructions in that documentation page don't show the grants as they're supposed to. Showing them for a specific database did work (but that doesn't feature in that doc page - it only says things like `SHOW GRANTS FOR db_admin;`). I never would have thought of that except for happening upon @rafiss comment here.

**Comment by knz** [comment]
@jazoom please followup here: https://github.com/cockroachdb/docs/issues/12967

**Comment by rafiss** [comment]
The behavior of `show grants for non_existent_user;` is fixed in v22.1 -- it will return an error saying `role/user "non_existent_user" does not exist`.
